engineering/frontend/code-style.md

```diff
@@ -0,0 +1,84 @@
+
+- Code should be functional in style rather than Object Orientated or Imperative unless there are no clean alternatives.
+
+  - Use pure functions where possible to make them testable and modular.
+  - Avoid mutating variables and the `let` keyword.
+  - React Components should be stateless functional components where possible.
+  - Avoid classes and stateful modules where possible.
+  - Don't Repeat Yourself. Make extensive use of the constants and utils files for re-usable strings and methods.
+  - The above rules can be relaxed for test scripts.
+
+- React components should be simple and composable and cater to real life UI design problems:
+
+  - **Simplicity**: Strive to keep the component API fairly simple and show real-world scenarios of using the component.
+  - **Representational**: React components should be templates, free of logic, and purely presentational. It aims to make our components shareable and easy to test.
+
+  - **Composition**: Break down components into smaller parts with minimal props to keep complexity low, and compose them together. This will ensure that the styles and functionality are flexible and extensible.
+
+  - **Accessibility**: When creating a component, keep accessibility top of mind. This includes keyboard navigation, focus management, color contrast, voice over, and the correct aria-\* attributes.
+
+  - **Naming Props**: We all know naming is the hardest thing in this industry. Generally, ensure a prop name is indicative of what it does. Boolean props should be named using auxiliary verbs such as does, has, is and should. For example, Button uses `isDisabled`, `isLoading`, etc.
+
+- When doing styling, strive to use `tailwindcss` classes for consistency and painless maintenance.
+
+  - Avoid hard coding colors. All colors from the design guideline should be pre-defined in [`tailwind.js`](../tailwind.config.js). Please use them with respect.
+
+  - Try to avoid injecting styles via `style` attribute because it's impossible to build responsiveness with inline styles. Strive to only use Tailwind classes for sizing, spacing, and building grid layouts.
+
+    ```jsx
+    // Don't
+    <div style={{marginLeft: '8px'}}></div
+
+    // Do
+    <div className="ml-2"></div
+
+    // Grid layout
+    <div className="grid grid-cols-3 gap-5">
+      <div className="col-span-2"></div>
+      <div className="col-span-1"></div>
+    </div>
+    ```
+
+- Maintain the separation of concerns in the folder, for example:
+
+  - `constants/` contains shared variables used across the app.
+  - `components/` contains shared ui components used across the app.
+  - `utils/` contains shared functions used across the app.
+  - `hooks/` contains shared hooks used across the app.
+  - `types/` contains common `TypeScript` types or interfaces.
```

Comment:
should we also include `contexts/` or `store/redux` for state management since all apps gonna need a state management

Reply:
I will add a section dedicated to state management to tech-ecosystem doc

Reply:
Added

> ## React Context
> We don't advocate for any state management library at the moment. One of the reasons libraries like Redux were so successful is because they solved the problem of [prop drilling](https://kentcdodds.com/blog/prop-drilling). However, the born of [Context API](https://reactjs.org/docs/context.html) made it easy to pass data through the component tree without having to pass props down manually at every level. More importantly, we think most applications don't need a complicated global state, and Context API is more than enough to solve that uncomplex problem. Our philosophy in React state management is simple: try to keep state as local as possible and use React context when prop drilling becomes a worry.